### PR TITLE
Set minimum supported C++ standard to 17

### PR DESCRIPTION
**Pull request by vepadulano**
Starting to include the changes needed to set the minimum supported c++ standard to 17. Once all the changes are recognized properly and the CI is green I will take care of improving the commit history.

### CODE REVIEW

**Comment by vepadulano** [comment]
@phsft-bot build with flags -DCMAKE_CXX_STANDARD=17

**Comment by hahnjo** [comment]
> Build failed on ROOT-ubuntu2204/nortcxxmod. Running on root-ubuntu-2204-2.cern.ch:/home/sftnight/build/workspace/root-pullrequests-build [See console output](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/179739/console).
> ### Failing tests:
> 
> * [projectroot.roottest.root.core.roottest_root_core_execStatusBitsCheck](https://lcgapp-services.cern.ch/root-jenkins/job/root-pullrequests-build/179739/testReport/projectroot.roottest.root/core/roottest_root_core_execStatusBitsCheck/)

This one is ~ known: https://github.com/root-project/root/issues/13058 (yes, that one *must* be solved before merging this PR). For `alma8`, I've added the "clean build" label to avoid incremental builds in the new CI.

You should probably also update `core/foundation/inc/ROOT/RConfig.hxx` to check for C++17. Other than that, this looks really good, other cleanups can be taken care of in future PRs as we go.

**Comment by vepadulano** [comment]
Still can't understand what is forcing `CMAKE_CXX_STANDARD=14` in the default PR builds... `rootspi` doesnt' explicitly set it, nor could I find it in `jenkins-pipelines`... still investigating

**Comment by vepadulano** [comment]
@phsft-bot build with flags -DCMAKE_CXX_STANDARD=17

**Comment by hahnjo** [comment]
> Still can't understand what is forcing `CMAKE_CXX_STANDARD=14` in the default PR builds... `rootspi` doesnt' explicitly set it, nor could I find it in `jenkins-pipelines`... still investigating

This *should* be https://github.com/root-project/rootspi/blob/master/jenkins/root-build.cmake#L931 , no? We likely need a new case
```groovy
      elseif (ROOT_VERSION VERSION_GREATER_EQUAL 6.29)
        set(specflags ${specflags} -DCMAKE_CXX_STANDARD=17)
      elseif (ROOT_VERSION VERSION_GREATER_EQUAL 6.25)
        set(specflags ${specflags} -DCMAKE_CXX_STANDARD=14)
      else()
        set(specflags ${specflags} -DCMAKE_CXX_STANDARD=11)
      endif()
```

**Comment by vepadulano** [comment]
> This should be https://github.com/root-project/rootspi/blob/master/jenkins/root-build.cmake#L931 , no? We likely need a new case

Yes, thank you so much! Seems like it, how could I not see that yesterday, I only focused on the `cxx14` flag and didn't see anything setting that xD Anyway, probably we want to merge this PR before adding that new flag to the PR builds, otherwise every other PR will build with C++17.

**Comment by hahnjo** [comment]
> Anyway, probably we want to merge this PR before adding that new flag to the PR builds, otherwise every other PR will build with C++17.

I think that would be fine though because it's decided that we go to C++14. IMHO losing that coverage before this PR is merged may be just fine and de-complicate things. It requires fixing https://github.com/root-project/root/issues/13058 first though; I can have another look in the afternoon, but this probably requires @pcanal to have a look...

**Comment by vepadulano** [comment]
@phsft-bot build with flags -DCMAKE_CXX_STANDARD=17

**Comment by vepadulano** [comment]
@phsft-bot build with flags -DCMAKE_CXX_STANDARD=17

**Comment by vepadulano** [comment]
@phsft-bot build

File: core/meta/test/CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 ROOT_ADD_GTEST(testStatusBitsChecker testStatusBitsChecker.cxx LIBRARIES Core)
 ROOT_ADD_GTEST(testHashRecursiveRemove testHashRecursiveRemove.cxx LIBRARIES Core)
-if(NOT MSVC OR win_broken_tests OR CMAKE_CXX_STANDARD GREATER 14)
+if(NOT MSVC OR win_broken_tests)
```
Comment (hahnjo) [inline core/meta/test/CMakeLists.txt:9]: I believe this means the test should be enabled for C++17, even on Windows? @bellenot

**Review by hahnjo — APPROVED** [review]
LGTM, the new CI seems mostly happy. We should wait for `ROOT-debian10-i386` and re-run `ROOT-ubuntu2204/nortcxxmod` on Jenkins (and hope for a different node than `root-ubuntu-2204-3`), just to be sure.

One minor nit inline, but can also be addressed in a follow-up. Let's get this in!

File: core/meta/test/CMakeLists.txt (hunk 1 above)
Comment (bellenot) [reply core/meta/test/CMakeLists.txt:9]: correct

**Comment by vepadulano** [comment]
All jenkins builds are green, except for ubuntu22 nortcxxmod since we landed on the wrong machine, and for debian10 which always takes too long. I will restart the build on ubuntu22 hoping to land on a working machine

**Comment by vepadulano** [comment]
@phsft-bot build just on ROOT-ubuntu2204/nortcxxmod, ROOT-debian10-i386/soversion

**Pull request merged**